File: engine/src/vulkan/pipeline.rs
use super::instance::VulkanInstance;
use super::render_pass::{self, VulkanRenderPass};
use super::{context::VulkanContext, device::VulkanDevice};
use anyhow::{Ok, Result};
use vulkanalia::bytecode::Bytecode;
use vulkanalia::vk::{self, DeviceV1_0, Handle, HasBuilder};

#[derive(Debug)]
pub struct VulkanPipeline {
    render_pass: VulkanRenderPass,
}

impl VulkanPipeline {
    pub unsafe fn create(
        instance: &VulkanInstance,
        device: &VulkanDevice,
        context: &mut VulkanContext,
    ) -> Result<VulkanPipeline> {
        // Render pass
        let render_pass = VulkanRenderPass::create(device, context)?;

        let vert = include_bytes!("../../../shaders/vert.spv");
        let frag = include_bytes!("../../../shaders/frag.spv");

        let vertex_shader_module = VulkanPipeline::create_shader_module(device, &vert[..])?;
        let fragment_shader_module = VulkanPipeline::create_shader_module(device, &frag[..])?;

        let vert_stage = vk::PipelineShaderStageCreateInfo::builder()
            .stage(vk::ShaderStageFlags::VERTEX)
            .module(vertex_shader_module)
            .name(b"main\0");

        let frag_stage = vk::PipelineShaderStageCreateInfo::builder()
            .stage(vk::ShaderStageFlags::FRAGMENT)
            .module(fragment_shader_module)
            .name(b"main\0");

        let vertex_input_state = vk::PipelineVertexInputStateCreateInfo::builder();
        let input_assembly_state = vk::PipelineInputAssemblyStateCreateInfo::builder()
            .topology(vk::PrimitiveTopology::TRIANGLE_LIST)
            .primitive_restart_enable(false);

        let viewport = vk::Viewport::builder()
            .x(0.0)
            .y(0.0)
            .width(context.swapchain_extent.width as f32)
            .height(context.swapchain_extent.height as f32)
            .min_depth(0.0)
            .max_depth(1.0);

        let scissor = vk::Rect2D::builder()
            .offset(vk::Offset2D { x: 0, y: 0 })
            .extent(context.swapchain_extent);

        let viewports = &[viewport];
        let scissors = &[scissor];
        let viewport_state = vk::PipelineViewportStateCreateInfo::builder()
            .viewports(viewports)
            .scissors(scissors);

        // rasterizer
        let rasterization_state = vk::PipelineRasterizationStateCreateInfo::builder()
            .depth_clamp_enable(false)
            .rasterizer_discard_enable(false)
            .polygon_mode(vk::PolygonMode::FILL)
            .line_width(1.0)
            .cull_mode(vk::CullModeFlags::BACK)
            .front_face(vk::FrontFace::CLOCKWISE)
            .depth_bias_enable(false);

        // multisampling
        let multisample_state = vk::PipelineMultisampleStateCreateInfo::builder()
            .sample_shading_enable(false)
            .rasterization_samples(vk::SampleCountFlags::_1);

        // color blending
        let attachment = vk::PipelineColorBlendAttachmentState::builder()
            .color_write_mask(vk::ColorComponentFlags::all())
            .blend_enable(true)
            .src_color_blend_factor(vk::BlendFactor::SRC_ALPHA)
            .dst_color_blend_factor(vk::BlendFactor::ONE_MINUS_SRC_ALPHA)
            .color_blend_op(vk::BlendOp::ADD)
            .src_alpha_blend_factor(vk::BlendFactor::ONE)
            .dst_alpha_blend_factor(vk::BlendFactor::ZERO)
            .alpha_blend_op(vk::BlendOp::ADD);

        let attachments = &[attachment];
        let color_blend_state = vk::PipelineColorBlendStateCreateInfo::builder()
            .logic_op_enable(false)
            .logic_op(vk::LogicOp::COPY)
            .attachments(attachments)
            .blend_constants([0.0, 0.0, 0.0, 0.0]);

        // layout
        let layout_info = vk::PipelineLayoutCreateInfo::builder();
        context.pipeline_layout = device
            .vk_device
            .create_pipeline_layout(&layout_info, None)?;

        let stages = &[vert_stage, frag_stage];
        let info = vk::GraphicsPipelineCreateInfo::builder()
            .stages(stages)
            .vertex_input_state(&vertex_input_state)
            .input_assembly_state(&input_assembly_state)
            .viewport_state(&viewport_state)
            .rasterization_state(&rasterization_state)
            .multisample_state(&multisample_state)
            .color_blend_state(&color_blend_state)
            .layout(context.pipeline_layout)
            .render_pass(context.render_pass)
            .subpass(0);

        context.pipeline = device
            .vk_device
            .create_graphics_pipelines(vk::PipelineCache::null(), &[info], None)?
            .0[0];

        // destroy shader modules
        device
            .vk_device
            .destroy_shader_module(vertex_shader_module, None);
        device
            .vk_device
            .destroy_shader_module(fragment_shader_module, None);

        Ok(VulkanPipeline { render_pass })
    }

    unsafe fn create_shader_module(
        device: &VulkanDevice,
        bytecode: &[u8],
    ) -> Result<vk::ShaderModule> {
        let bytecode = Bytecode::new(bytecode).unwrap();
        let info = vk::ShaderModuleCreateInfo::builder()
            .code_size(bytecode.code_size())
            .code(bytecode.code());

        Ok(device.vk_device.create_shader_module(&info, None)?)
    }

    pub unsafe fn destroy(&mut self, device: &VulkanDevice, context: &mut VulkanContext) {
        device
            .vk_device
            .destroy_command_pool(context.command_pool, None);
        context
            .framebuffers
            .iter()
            .for_each(|f| device.vk_device.destroy_framebuffer(*f, None));
        device.vk_device.destroy_pipeline(context.pipeline, None);
        device
            .vk_device
            .destroy_pipeline_layout(context.pipeline_layout, None);
        device
            .vk_device
            .destroy_render_pass(context.render_pass, None);
    }
}

File: engine/renderer/vulkan/pipeline.go
package vulkan

import (
	"fmt"

	vk "github.com/goki/vulkan"
	"github.com/spaghettifunk/anima/engine/core"
	"github.com/spaghettifunk/anima/engine/renderer/metadata"
)

/**
 * @brief Holds a Vulkan pipeline and its layout.
 */
type VulkanPipeline struct {
	/** @brief The internal pipeline handle. */
	Handle vk.Pipeline
	/** @brief The pipeline layout. */
	PipelineLayout vk.PipelineLayout
}

type VulkanPipelineConfig struct {
	/** @brief A pointer to the renderpass to associate with the pipeline. */
	Renderpass *VulkanRenderPass
	/** @brief The stride of the vertex data to be used (ex: sizeof(vertex_3d)) */
	Stride uint32
	/** @brief An array of attributes. */
	Attributes []vk.VertexInputAttributeDescription
	/** @brief An array of descriptor set layouts. */
	DescriptorSetLayouts []vk.DescriptorSetLayout
	/** @brief An VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BITarray of stages. */
	Stages []vk.PipelineShaderStageCreateInfo
	/** @brief The initial viewport configuration. */
	Viewport vk.Viewport
	/** @brief The initial scissor configuration. */
	Scissor vk.Rect2D
	/** @brief The face cull mode. */
	CullMode metadata.FaceCullMode
	/** @brief Indicates if this pipeline should use wireframe mode. */
	IsWireframe bool
	/** @brief The shader flags used for creating the pipeline. */
	ShaderFlags metadata.ShaderFlagBits
	/** @brief An array of push constant data ranges. */
	PushConstantRanges []*metadata.MemoryRange
}

func NewGraphicsPipeline(context *VulkanContext, config *VulkanPipelineConfig) (*VulkanPipeline, error) {
	outPipeline := &VulkanPipeline{}

	// Viewport state
	viewportState := vk.PipelineViewportStateCreateInfo{
		SType:         vk.StructureTypePipelineViewportStateCreateInfo,
		ViewportCount: 1,
		PViewports:    []vk.Viewport{config.Viewport},
		ScissorCount:  1,
		PScissors:     []vk.Rect2D{config.Scissor},
	}
	viewportState.Deref()

	// Rasterizer
	rasterizerCreateInfo := vk.PipelineRasterizationStateCreateInfo{
		SType:                   vk.StructureTypePipelineRasterizationStateCreateInfo,
		DepthClampEnable:        vk.False,
		RasterizerDiscardEnable: vk.False,
		PolygonMode:             vk.PolygonModeLine,
		LineWidth:               1.0,
		FrontFace:               vk.FrontFaceCounterClockwise,
		DepthBiasEnable:         vk.False,
		DepthBiasConstantFactor: 0.0,
		DepthBiasClamp:          0.0,
		DepthBiasSlopeFactor:    0.0,
	}
	if !config.IsWireframe {
		rasterizerCreateInfo.PolygonMode = vk.PolygonModeFill
	}
	switch config.CullMode {
	case metadata.FaceCullModeNone:
		rasterizerCreateInfo.CullMode = vk.CullModeFlags(vk.CullModeNone)
	case metadata.FaceCullModeFront:
		rasterizerCreateInfo.CullMode = vk.CullModeFlags(vk.CullModeFrontBit)
	case metadata.FaceCullModeFrontAndBack:
		rasterizerCreateInfo.CullMode = vk.CullModeFlags(vk.CullModeFrontAndBack)
	default:
		fallthrough
	case metadata.FaceCullModeBack:
		rasterizerCreateInfo.CullMode = vk.CullModeFlags(vk.CullModeBackBit)
	}
	rasterizerCreateInfo.Deref()

	// Multisampling.
	multisamplingCreateInfo := vk.PipelineMultisampleStateCreateInfo{
		SType:                 vk.StructureTypePipelineMultisampleStateCreateInfo,
		SampleShadingEnable:   vk.False,
		RasterizationSamples:  vk.SampleCount1Bit,
		MinSampleShading:      1.0,
		PSampleMask:           nil,
		AlphaToCoverageEnable: vk.False,
		AlphaToOneEnable:      vk.False,
	}
	multisamplingCreateInfo.Deref()

	// Depth and stencil testing.
	depthStencil := vk.PipelineDepthStencilStateCreateInfo{
		SType:             vk.StructureTypePipelineDepthStencilStateCreateInfo,
		DepthTestEnable:   vk.False,
		DepthWriteEnable:  vk.False,
		StencilTestEnable: vk.False,
	}
	if (metadata.ShaderFlags(config.ShaderFlags) & metadata.SHADER_FLAG_DEPTH_TEST) != 0 {
		depthStencil.DepthTestEnable = vk.True
		depthStencil.DepthCompareOp = vk.CompareOpLess
		depthStencil.DepthBoundsTestEnable = vk.False
		depthStencil.StencilTestEnable = vk.False
	}
	if (metadata.ShaderFlags(config.ShaderFlags) & metadata.SHADER_FLAG_DEPTH_WRITE) != 0 {
		depthStencil.DepthWriteEnable = vk.True
	}
	depthStencil.Deref()

	colorBlendAttachmentState := vk.PipelineColorBlendAttachmentState{
		BlendEnable:         vk.True,
		SrcColorBlendFactor: vk.BlendFactorSrcAlpha,
		DstColorBlendFactor: vk.BlendFactorOneMinusSrcAlpha,
		ColorBlendOp:        vk.BlendOpAdd,
		SrcAlphaBlendFactor: vk.BlendFactorSrcAlpha,
		DstAlphaBlendFactor: vk.BlendFactorOneMinusSrcAlpha,
		AlphaBlendOp:        vk.BlendOpAdd,
		ColorWriteMask: vk.ColorComponentFlags(vk.ColorComponentRBit) | vk.ColorComponentFlags(vk.ColorComponentGBit) |
			vk.ColorComponentFlags(vk.ColorComponentBBit) | vk.ColorComponentFlags(vk.ColorComponentABit),
	}
	colorBlendAttachmentState.Deref()

	colorBlendStateCreateInfo := vk.PipelineColorBlendStateCreateInfo{
		SType:           vk.StructureTypePipelineColorBlendStateCreateInfo,
		LogicOpEnable:   vk.False,
		LogicOp:         vk.LogicOpCopy,
		AttachmentCount: 1,
		PAttachments:    []vk.PipelineColorBlendAttachmentState{colorBlendAttachmentState},
	}
	colorBlendStateCreateInfo.Deref()

	// Dynamic state
	dynamicStates := []vk.DynamicState{
		vk.DynamicStateViewport,
		vk.DynamicStateScissor,
		vk.DynamicStateLineWidth,
	}

	dynamicStateCreateInfo := vk.PipelineDynamicStateCreateInfo{
		SType:             vk.StructureTypePipelineDynamicStateCreateInfo,
		DynamicStateCount: uint32(len(dynamicStates)),
		PDynamicStates:    dynamicStates,
	}
	dynamicStateCreateInfo.Deref()

	// Vertex input
	bindingDescription := vk.VertexInputBindingDescription{
		Binding:   0, // Binding index
		Stride:    config.Stride,
		InputRate: vk.VertexInputRateVertex, // Move to next data entry for each vertex.
	}
	bindingDescription.Deref()

	// Attributes
	vertexInputInfo := vk.PipelineVertexInputStateCreateInfo{
		SType:                           vk.StructureTypePipelineVertexInputStateCreateInfo,
		VertexBindingDescriptionCount:   1,
		PVertexBindingDescriptions:      []vk.VertexInputBindingDescription{bindingDescription},
		VertexAttributeDescriptionCount: uint32(len(config.Attributes)),
		PVertexAttributeDescriptions:    config.Attributes,
	}
	vertexInputInfo.Deref()

	// Input assembly
	inputAssembly := vk.PipelineInputAssemblyStateCreateInfo{
		SType:                  vk.StructureTypePipelineInputAssemblyStateCreateInfo,
		Topology:               vk.PrimitiveTopologyTriangleList,
		PrimitiveRestartEnable: vk.False,
	}
	inputAssembly.Deref()

	// Pipeline layout
	pipelineLayoutCreateInfo := vk.PipelineLayoutCreateInfo{
		SType:                  vk.StructureTypePipelineLayoutCreateInfo,
		SetLayoutCount:         uint32(len(config.DescriptorSetLayouts)),
		PSetLayouts:            config.DescriptorSetLayouts,
		PushConstantRangeCount: 0,
		PPushConstantRanges:    nil,
	}

	// Push constants
	if len(config.PushConstantRanges) > 0 {
		if len(config.PushConstantRanges) > 32 {
			err := fmt.Errorf("func NewGraphicsPipeline: cannot have more than 32 push constant ranges. Passed count: %d", len(config.PushConstantRanges))
			return nil, err
		}

		// NOTE: 32 is the max number of ranges we can ever have, since spec only guarantees 128 bytes with 4-byte alignment.
		ranges := make([]vk.PushConstantRange, 32)
		for i := 0; i < len(config.PushConstantRanges); i++ {
			ranges[i].StageFlags = vk.ShaderStageFlags(vk.ShaderStageVertexBit) | vk.ShaderStageFlags(vk.ShaderStageFragmentBit)
			ranges[i].Offset = uint32(config.PushConstantRanges[i].Offset)
			ranges[i].Size = uint32(config.PushConstantRanges[i].Size)
			ranges[i].Deref()
		}
		pipelineLayoutCreateInfo.PushConstantRangeCount = uint32(len(config.PushConstantRanges))
		pipelineLayoutCreateInfo.PPushConstantRanges = ranges
	}
	pipelineLayoutCreateInfo.Deref()

	// Create the pipeline layout.
	var pPipelineLayout vk.PipelineLayout

	if err := lockPool.SafeCall(PipelineManagement, func() error {
		result := vk.CreatePipelineLayout(
			context.Device.LogicalDevice,
			&pipelineLayoutCreateInfo,
			context.Allocator,
			&pPipelineLayout)
		if !VulkanResultIsSuccess(result) {
			err := fmt.Errorf("vkCreatePipelineLayout failed with %s", VulkanResultString(result, true))
			return err
		}
		outPipeline.PipelineLayout = pPipelineLayout
		return nil
	}); err != nil {
		return nil, err
	}

	// Pipeline create
	pipelineCreateInfo := vk.GraphicsPipelineCreateInfo{
		SType:               vk.StructureTypeGraphicsPipelineCreateInfo,
		StageCount:          uint32(len(config.Stages)),
		PStages:             config.Stages,
		PVertexInputState:   &vertexInputInfo,
		PInputAssemblyState: &inputAssembly,
		PViewportState:      &viewportState,
		PRasterizationState: &rasterizerCreateInfo,
		PMultisampleState:   &multisamplingCreateInfo,
		PDepthStencilState:  &depthStencil,
		PColorBlendState:    &colorBlendStateCreateInfo,
		PDynamicState:       &dynamicStateCreateInfo,
		PTessellationState:  nil,
		Layout:              outPipeline.PipelineLayout,
		RenderPass:          config.Renderpass.Handle,
		Subpass:             0,
		BasePipelineHandle:  vk.NullPipeline,
		BasePipelineIndex:   -1,
	}
	pipelineCreateInfo.Deref()

	pPipelines := []vk.Pipeline{outPipeline.Handle}

	if err := lockPool.SafeCall(PipelineManagement, func() error {
		result := vk.CreateGraphicsPipelines(
			context.Device.LogicalDevice,
			vk.NullPipelineCache,
			1,
			[]vk.GraphicsPipelineCreateInfo{pipelineCreateInfo},
			context.Allocator,
			pPipelines)

		if !VulkanResultIsSuccess(result) {
			err := fmt.Errorf("vkCreateGraphicsPipelines failed with %s", VulkanResultString(result, true))
			return err
		}
		return nil
	}); err != nil {
		return nil, err
	}

	if len(pPipelines) <= 0 {
		err := fmt.Errorf("vulkan pipeline handle is nil")
		return nil, err
	}

	outPipeline.Handle = pPipelines[0]

	core.LogDebug("Graphics pipeline created!")
	return outPipeline, nil
}

func (pipeline *VulkanPipeline) Destroy(context *VulkanContext) error {
	// Destroy pipeline
	if pipeline.Handle != nil {
		if err := lockPool.SafeCall(PipelineManagement, func() error {
			vk.DestroyPipeline(context.Device.LogicalDevice, pipeline.Handle, context.Allocator)
			pipeline.Handle = nil
			return nil
		}); err != nil {
			return err
		}
	}
	// Destroy layout
	if pipeline.PipelineLayout != nil {
		if err := lockPool.SafeCall(PipelineManagement, func() error {
			vk.DestroyPipelineLayout(context.Device.LogicalDevice, pipeline.PipelineLayout, context.Allocator)
			pipeline.PipelineLayout = nil
			return nil
		}); err != nil {
			return err
		}
	}
	return nil
}

func (pipeline *VulkanPipeline) Bind(command_buffer *VulkanCommandBuffer, bind_point vk.PipelineBindPoint) error {
	if err := lockPool.SafeCall(CommandBufferManagement, func() error {
		vk.CmdBindPipeline(command_buffer.Handle, bind_point, pipeline.Handle)
		return nil
	}); err != nil {
		return err
	}
	return nil
}
